src/iterators.jl

```diff
@@ -0,0 +1,179 @@
+"""
+  abstract type AbstractNetworkIterator
+
+A stateful iterator with two states: `increment!` and `compute!`. Each iteration begins
+with a call to `increment!` before executing `compute!`, however the initial call to
+`iterate` skips the `increment!` call as it is assumed the iterator is initalized such that 
+this call is implict. Termination of the iterator is controlled by the function `done`.
+"""
+abstract type AbstractNetworkIterator end
+
+# We use greater than or equals here as we increment the state at the start of the iteration
+islaststep(iterator::AbstractNetworkIterator) = state(iterator) >= length(iterator)
+
+function Base.iterate(iterator::AbstractNetworkIterator, init = true)
+    # The assumption is that first "increment!" is implicit, therefore we must skip the
+    # the termination check for the first iteration, i.e. `AbstractNetworkIterator` is not
+    # defined when length < 1,
+    init || islaststep(iterator) && return nothing
+    # We seperate increment! from step! and demand that any AbstractNetworkIterator *must*
+    # define a method for increment! This way we avoid cases where one may wish to nest
+    # calls to different step! methods accidentaly incrementing multiple times.
+    init || increment!(iterator)
+    rv = compute!(iterator)
+    return rv, false
+end
+
+increment!(iterator::AbstractNetworkIterator) = throw(MethodError(increment!, Tuple{typeof(iterator)}))
+compute!(iterator::AbstractNetworkIterator) = iterator
+
+step!(iterator::AbstractNetworkIterator) = step!(identity, iterator)
+function step!(f, iterator::AbstractNetworkIterator)
+    compute!(iterator)
+    f(iterator)
+    increment!(iterator)
+    return iterator
+end
+
+#
+# RegionIterator
+#
+"""
+  struct RegionIterator{Problem, RegionPlan} <: AbstractNetworkIterator
+"""
+mutable struct RegionIterator{Problem, RegionPlan} <: AbstractNetworkIterator
+    problem::Problem
+    region_plan::RegionPlan
+    which_region::Int
+    const which_sweep::Int
+    function RegionIterator(problem::P, region_plan::R, sweep::Int) where {P, R}
+        try
+            first(region_plan)
+        catch e
+            if e isa BoundsError
+                throw(ArgumentError("Cannot construct a region iterator with 0 elements."))
+            end
+            rethrow()
+        end
```

Comment:
I'm always a bit weary to use `try` except in extenuating circumstances, when is it not enough to use something like `isempty(region_plan)`? It seems reasonable to expect whatever is input as `region_plan` to define some functions like `isempty`, `length`, etc.

Reply:
Yes I think this would be easier to discuss in person, I've gotten a little bit lost around those edge cases in the design. If you want to write a custom error message I don't see why something closer to the previous design doesn't suffice, i.e.:
```julia
        if isempty(region_plan)
            throw(ArgumentError("Cannot construct a region iterator with 0 elements."))
        end
```
but maybe I'm missing something.